File: 314-binary-tree-vertical-order-traversal/binary-tree-vertical-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<List<Integer>> res = new LinkedList<>();
    HashMap<Integer, List<Integer>> map = new HashMap<>();
    public List<List<Integer>> verticalOrder(TreeNode root) {
        helper(root);
        int[] keys = new int[map.keySet().size()];
        int idx = 0;
        for (int key : map.keySet()) {
            keys[idx++] = key;
        }
        Arrays.sort(keys);
        for (int key : keys) {
            res.add(map.get(key));
        }
        return res;
    }

    public void helper(TreeNode root) {
        Queue<Pair<TreeNode, Integer>> queue = new LinkedList<>();
        if (root != null) {
            queue.add(new Pair(root, 0));
        }
        while (!queue.isEmpty()) {
            Pair<TreeNode, Integer> p = queue.poll();
            TreeNode node = p.getKey();
            int n = p.getValue();
            List<Integer> list = map.getOrDefault(n, new LinkedList<Integer>());
            list.add(node.val);
            map.put(n, list);
            if (node.left != null) {
                queue.add(new Pair(node.left, n - 1));
            }
            if (node.right != null) {
                queue.add(new Pair(node.right, n + 1));
            }
        }
    }
}

File: 314-binary-tree-vertical-order-traversal/binary-tree-vertical-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        map = defaultdict(list)
        queue = deque()
        if root:
            queue.append((root, 0))
        while queue:
            size  = len(queue)
            # get all nodes in current level
            for _ in range(size):
                node, idx = queue.popleft()
                # add node to map based on idx
                map[idx].append(node.val)
                # add left child node and right child node
                if node.left:
                    queue.append((node.left, idx - 1))
                if node.right:
                    queue.append((node.right, idx + 1))
        keys = list(map.keys())
        keys.sort()
        return [map[k] for k in keys]